keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,916 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this proposal, the Kubernetes API server will enable per-request read/write
+timeouts using the `ResponseController` functionality provided by the golang net/http library.
+With the enablement of per-request read/write timeout, the `timeout` filter in the
+kube-apiserver will become redundant, and thus be removed from the server handler chain.
+
+
+## Motivation
+
+The `timeout` filter has been a source of many hard to debug data races in the past. 
+With the removal of the `timeout` filter from the apiserver handler chain, the entire
+request handler will be executed on a single goroutine. Thus it will eliminate 
+the data race conditions that are inherent with the `timeout` filter.
+
+### Goals
+
+The primary goal is to reduce data race conditions by removing the `timeout`
+filter from the apiserver request handler chain.
+
+### Non-Goals
+
+- WATCH requests are outside the scope of this proposal, since currently we
+don't attach any deadline to the `Context` of a WATCH request. An orthogonal
+effort to allocate a deadline to the `Context`of WATCH request(s) is a
+prerequisite, and such effort is outside the scope of this proposal.
+
+- All other long-running requests are also outside the scope of this proposal.
+
+
+## Proposal
+
+A new feature flag `PerHandlerReadWriteTimeout` will be introduced, when enabled:
+- On start-up, the kube-apiserver will remove the `timeout` filter
+  from the request handler chain.
+- To every incoming non long-running request, the apiserver will 
+  apply a read/write deadline.
+- Replace the asynchronous mutating-request executor with one that 
+  executes synchronously, in the same goroutine as the caller.
+  
+We also need to maintain parity with logs/metrics that are produced
+by the `timeout` filter today for any request that times out.
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+##### Client Hanging Indefinitely
+
+With the `timeout` filter enabled today, it sends a `504 GatewayTimeout` response
+or an `error` to the client, if the request handler it is executing is blocking indefinitely.
+
+If the client does not enforce any timeout on its side, for example, not setting a
+global `Client` timeout, or not setting a deadline to the request `Context`; the client
+can hang forever waiting for a response from the server. This is certainly true for
+HTTP/1x where [net/http does not execute the request handler in a new goroutine](https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L2031-L2039).
+
+> It's worth noting that by default `client-go` does not enforce any timeout globally.
+
+With the removal of the `timeout` filter, we need to ensure that a request handler that
+blocks on the server should never block a client that does not enforce timeout on its side.
```

Comment:
How exactly you're going to ensure it?

Reply:
I have opened an issue with the golang team, to understand whether this is expected. https://github.com/golang/go/issues/65526